### PR TITLE
check job ActiveDeadlineSeconds

### LINKED ISSUE #32149: Too long time needed for a job to be killed after exceeding relatively short ActiveDeadlineSeconds
If you set relatively short `.spec.activeDeadlineSeconds`, let's say 60 seconds (or generally anything less than full resync time which is 10 mins) and your job stabilizes after few initial seconds, iow. now new pods are created or anything involving job controller, you end up waiting more than 10 minutes before the Job is actually killed due to exceeding ADS. 
The reason for that is that controller synchronizes objects only when job itself or underlying pod changes. Or, when full resync is performed once 10 mins. This leads to unnecessarily wait that longer timeout. 

Possible fixes:
1. Mark jobs with ADS set and have them full resynced more often.
2. Queue jobs with ADS set, so they are resynced more often.
   
   Other possibilities?

@erictune @janetkuo ideas?

### PR DESCRIPTION
**Pull request by weiwei04**
**What this PR does / why we need it**:

enqueue a sync task after ActiveDeadlineSeconds

**Which issue this PR fixes** *: 

fixes #32149

**Special notes for your reviewer**:

**Release note**:

```release-note
enqueue a sync task to wake up jobcontroller to check job ActiveDeadlineSeconds in time
```

### CODE REVIEW

**Comment by weiwei04** [comment]
@janetkuo  please take a look, thanks.

**Comment by soltysh** [comment]
@weiwei04 sorry, was out for quite a long, will look into it today or tomorrow.

**Comment by weiwei04** [comment]
@soltysh thanks:)

File: pkg/controller/job/jobcontroller.go [hunk 1]
```diff
@@ -306,6 +302,32 @@ func (jm *JobController) deletePod(obj interface{}) {
 	jm.enqueueController(job)
 }
 
+func (jm *JobController) updateJob(old, cur interface{}) {
+	oldJob := cur.(*batch.Job)
+	curJob := cur.(*batch.Job)
+
+	// never return error
+	key, _ := controller.KeyFunc(curJob)
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: There's no point in re-trying if there was an error. Check error and just return if that happens.

File: pkg/controller/job/jobcontroller.go [hunk 2]
```diff
@@ -306,6 +302,32 @@ func (jm *JobController) deletePod(obj interface{}) {
...
+	jm.queue.Add(key)
+	// check if need to add a new rsync for ActiveDeadlineSeconds
+	if curJob.Status.StartTime != nil {
+		curSeconds := curJob.Spec.ActiveDeadlineSeconds
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: Can you please use `currentADS` and `oldADS` it's less confusing name :)

File: staging/src/k8s.io/api/batch/v1/types.go [hunk 3]
```diff
@@ -150,6 +150,10 @@ const (
 	JobFailed JobConditionType = "Failed"
 )
 
+const (
```
Comment (soltysh) [inline staging/src/k8s.io/api/batch/v1/types.go]: Not needed, drop it. It'll cause only confusion wrt to our API.

File: test/e2e/apps/job.go [hunk 4]
```diff
@@ -28,6 +28,8 @@ import (
 	"k8s.io/kubernetes/pkg/kubectl"
 	"k8s.io/kubernetes/test/e2e/framework"
 
+	batch "k8s.io/api/batch/v1"
```
Comment (soltysh) [inline test/e2e/apps/job.go]: Use `batchv1` as the name alias, to make it clear it's that version.

File: test/e2e/apps/job.go [hunk 5]
```diff
@@ -86,9 +88,20 @@ var _ = SIGDescribe("Job", func() {
 		Expect(err).NotTo(HaveOccurred())
 	})
 
+	It("should exceed active deadline", func() {
+		By("Creating a job")
+		var activeDeadlineSeconds int64 = 10
```
Comment (soltysh) [inline test/e2e/apps/job.go]: Make it shorter, this will cause the entire test suite longer by 10 seconds. I'd propose 1-2 seconds max.

**Review by soltysh — CHANGES_REQUESTED** [review]
Mostly nits, overall looks good.

File: pkg/controller/job/jobcontroller.go (hunk 2 above)
Comment (weiwei04) [inline pkg/controller/job/jobcontroller.go]: yes ADS would be a more meaningful name:)

File: test/e2e/apps/job.go (hunk 5 above)
Comment (weiwei04) [inline test/e2e/apps/job.go]: changed to 2 seconds.

**Comment by weiwei04** [comment]
@soltysh modified code according to your comments, please take a look, thanks:)

**Comment by weiwei04** [comment]
@soltysh modified code according to your comments, and all the test are passed, please take a look, thanks:)

**Comment by weiwei04** [comment]
ping @soltysh :)

**Comment by weiwei04** [comment]
ping @soltysh :)

**Comment by weiwei04** [comment]
ping @soltysh :)

**Comment by soltysh** [comment]
@weiwei04 sorry for the long silence, looking at it right now...

**Review by soltysh — APPROVED** [review]
LGTM

**Comment by soltysh** [comment]
@weiwei04 mind squashing these changes into a single commit and I'll apply all the necessary labels for merge

**Comment by weiwei04** [comment]
Thanks:)

**Pull request merged**